### PR TITLE
Catch job claim failures, and result in job termination

### PR DESCRIPTION
**Pull request by seanstory**
## Part of https://github.com/elastic/connectors/issues/2604

If a connector thinks a job is running for it, we'd better ensure that a job either starts or fails. Before this PR, if `sync_job.claim()` failed, the connector would not be informed that there was nothing running the job, and the job it expected to start running would never transition out of the pending state. This PR does a few things:

1. Ensures that after the connector record has been informed that the sync is starting, that everything past that is in a `try/except` block to handle any errors and fail the job.
2. wraps all SyncJob methods individually in `try/except` to ensure that any surfaced error messages are as actionable as possible
3. Adds tests

This PR does not yet address the root issue of the linked ticket, which is _why_ we're getting to a place where `claim()` is resulting in ES throwing a 413 error.

#### Pre-Review Checklist
- [x] this PR does NOT contain credentials of any kind, such as API keys or username/passwords (double check `config.yml.example`)
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

## Release Note

Fixes a bug where sync jobs can be permanently  stuck in "pending" (until manually canceled) because of an unfortunately timed Elasticsearch exception. Now such exceptions will cause jobs to fail and provide actionable error messaging.

### CODE REVIEW

File: connectors/sync_job_runner.py [hunk 1]
```diff
@@ -216,6 +217,7 @@ async def execute(self):
             self.sync_job.log_error(e, exc_info=True)
             await self._sync_done(sync_status=JobStatus.ERROR, sync_error=e)
         finally:
+            self.sync_job.log_info("Terminating the job in error, and cleaning up")
```
Comment (jedrazb) [inline connectors/sync_job_runner.py:220]: Afaik the `finally` block is always executed, regardless of whether an exception was raised or not. So this log might be misleading. We are just cleaning up here

Out of scope: We are already terminating sync in `except` clauses above this line with `self._sync_done(sync_status=JobStatus.ERROR ....` perhaps we could add logs in each of the clause (e.g. we don't log anything if job is suspended / canceled)

File: connectors/sync_job_runner.py [hunk 2]
```diff
@@ -374,7 +376,9 @@ async def _sync_done(self, sync_status, sync_error=None):
 
         if await self.reload_connector():
             sync_cursor = (
-                self.data_provider.sync_cursor()
+                None
+                if not self.data_provider  # If we failed before initializing the data provider, we don't need to change the cursor
```
Comment (jedrazb) [inline connectors/sync_job_runner.py:380]: Could the `None` sync_cursor cause problems down the line if there are intermittent issues with data provider initialization. 

I'm thinking of:
- we have connector, we run incremental syncs fine
- we once run into `if not self.data_provider` is `true` case
- we set this `None` cursor in connector doc `self.connector.sync_done(...`

How to recover for further incremental syncs (e.g. scheduled)? Can our logic communicate that user needs to run a full sync first given that the cursor is now `None`?

**Review by jedrazb — COMMENTED** [review]
Great improvement to the framework! 👍  I left some comments

File: connectors/sync_job_runner.py (hunk 1 above)
Comment (seanstory) [reply connectors/sync_job_runner.py:220]: omg you're totally right. I'd originally just added this as a "made it here" debugging log. I think I'll still leave something here for that purpose, but you're right, it shouldn't assume error state.

I can add extra logs too, that's easy.

File: connectors/sync_job_runner.py (hunk 2 above)
Comment (seanstory) [reply connectors/sync_job_runner.py:380]: Sync cursor only gets set if if the job completed successfully: 
https://github.com/elastic/connectors/blob/main/connectors/protocol/connectors.py#L763-L764 

so we shouldn't be at risk here. If `self.data_provider` is not set, we didn't complete successfully.

**Review by jedrazb — APPROVED** [review]
LGTM!

**Comment by seanstory** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.13|[](https://github.com/elastic/connectors/pull/2620)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**